File: God Zuo/Python/class054 单调队列/My02.py
#测试链接 : https://leetcode.cn/problems/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/
class Solution:
    def longestSubarray(self, nums: list[int], limit: int) -> int:
        n=len(nums)
        max_q = [0] * n
        min_q = [0] * n
        max_h = max_t = 0
        min_h = min_t = 0
        ans = 0
        r = 0
        def ok(limit,curr):
            max_val = nums[max_q[max_h]] if max_h < max_t and nums[max_q[max_h]] > curr else curr
            min_val = nums[min_q[min_h]] if min_h < min_t and nums[min_q[min_h]] < curr else curr
            return max_val - min_val <= limit
        
        def push(r):
            nonlocal max_t,min_t
            while max_h<max_t and nums[max_q[max_t-1]]<=nums[r]:
                max_t-=1
            max_q[max_t]=r
            max_t+=1
            while min_h<min_t and nums[min_q[min_t-1]]>=nums[r]:
                min_t-=1
            min_q[min_t]=r
            min_t+=1
        
        def pop(l):
            nonlocal max_h,min_h
            if max_h<max_t and max_q[max_h]==l:
                max_h+=1
            if min_h<min_t and min_q[min_h]==l:
                min_h+=1

        for l in range(n):
            while r<n and ok(limit,nums[r]):
                push(r)
                r+=1
            length=r-l
            ans=length if length>ans else ans 
            pop(l)
        
        return ans

File: God Zuo/class083 动态规划优化枚举下/My02.cpp
// K个逆序对数组
// 逆序对的定义如下：
// 对于数组nums的第i个和第j个元素
// 如果满足0<=i<j<nums.length 且 nums[i]>nums[j]，则为一个逆序对
// 给你两个整数n和k，找出所有包含从1到n的数字
// 且恰好拥有k个逆序对的不同的数组的个数
// 由于答案可能很大，答案对 1000000007 取模
// 测试链接 : https://leetcode.cn/problems/k-inverse-pairs-array/

#include<bits/stdc++.h>
using namespace std;

int MOD = 1e9 + 7;
int dp[1005][1005];

class Solution {
public:
    // dp[i][j]表示 1,2,3...i 这些数中又j个逆序对的排列数
    int kInversePairs(int n, int k) {
        for (int i = 0; i <= n; ++i){
            dp[i][0] = 1;
        }
        for (int i = 1; i <= n; ++i){
            int wd = 1;
            for(int j = 1; j <= k; ++j){
                if(i>j){
                    wd += dp[i - 1][j];
                    wd %= MOD;
                }else{
                    wd = ((wd + dp[i - 1][j]) % MOD - dp[i - 1][j - i] + MOD) % MOD;
                }
                dp[i][j] = wd;
            }
        }
        return dp[n][k];
    }
};